### PR TITLE
RFC: Add simple callbacks to cvode

### PR DESCRIPTION
**Pull request by lstagner**
This PR adds a simple callback that returns a Boolean indicated whether to continue the calculation. I am using this in my own [work](https://github.com/lstagner/GuidingCenterOrbits.jl/blob/master/src/orbit.jl#L171). 

I don't expect this to be merged but I would like to get peoples opinion on the approach taken and the general need of such a feature.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
I happen to be working on callbacks + event handling as outlined in JuliaDiffEq/DifferentialEquations.jl#64. I'm benchmarking and perfecting the implementation right now, but might be putting something like this into the Sundials simplified interface so that way the DifferentialEquations.jl Sundials wrapper will work seamlessly as well. However, since the Sundials' non-stiff Adams solver benchmarks as not being very fast/efficient in the first place in comparison to RK methods, I'll be taking my time on the Sundials side since it's not really too useful (only the BDF method seems to be efficient, but that's only for specific problems of course).

That said, the full interface will at least need enough to do the interpolation and root finding. The event handling should have a way of using the interpolation to hone in on the correct event time/value. Then you need to pull-back the cached "current value" since for a discontinuity you don't want to use those.

File: src/simple.jl [hunk 1]
```diff
@@ -121,12 +122,16 @@ function cvode(f::Function, y0::Vector{Float64}, t::Vector{Float64}, userdata::A
         tout = [0.0]
         for k in 2:length(t)
             flag = @checkflag CVode(mem, t[k], ynv, tout, CV_NORMAL)
+            if !callback(ynv)
```
Comment (ChrisRackauckas) [inline src/simple.jl]: This is a very specific type of callback. Maybe generalize it a bit?

Comment (lstagner) [inline src/simple.jl]: Generalize in what way? With clever uses of closures you can already accomplish quite a bit with this implementation.

Comment (ChrisRackauckas) [inline src/simple.jl]: Maybe give it the current time `t[k]` and the `mem` to allow interpolations (that's all that should be needed for `CVodeGetDky`, right?)? And add it to the docstring.

**Comment by ChrisRackauckas** [comment]
Tests on master v0.6 are failing due to unrelated reasons.

**Pull request merged**